FATAL error when trying to write large (> 8k) Portable object to map.

03/31 23:46:19:486 FATAL [GetOperation] [10.109.139.94]:6702 [prod-sync] [3.2] java.lang.IllegalArgumentException
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.IllegalArgumentException
        at com.hazelcast.nio.serialization.SerializationServiceImpl.handleException(SerializationServiceImpl.java:298)
        at com.hazelcast.nio.serialization.SerializationServiceImpl.toData(SerializationServiceImpl.java:194)
        at com.hazelcast.map.record.DataRecordFactory.newRecord(DataRecordFactory.java:49)
        at com.hazelcast.map.MapService.createRecord(MapService.java:410)
        at com.hazelcast.map.MapService.createRecord(MapService.java:405)
        at com.hazelcast.map.DefaultRecordStore.get(DefaultRecordStore.java:543)
        at com.hazelcast.map.operation.GetOperation.run(GetOperation.java:37)
        at com.hazelcast.spi.impl.BasicOperationService.processOperation(BasicOperationService.java:363)
        at com.hazelcast.spi.impl.BasicOperationService.access$300(BasicOperationService.java:102)
        at com.hazelcast.spi.impl.BasicOperationService$BasicOperationProcessorImpl.process(BasicOperationService.java:754)
        at com.hazelcast.spi.impl.BasicOperationScheduler$PartitionThread.process(BasicOperationScheduler.java:276)
        at com.hazelcast.spi.impl.BasicOperationScheduler$PartitionThread.doRun(BasicOperationScheduler.java:270)
        at com.hazelcast.spi.impl.BasicOperationScheduler$PartitionThread.run(BasicOperationScheduler.java:245)
Caused by: java.lang.IllegalArgumentException
        at com.hazelcast.nio.serialization.ByteArrayObjectDataOutput.position(ByteArrayObjectDataOutput.java:274)
        at com.hazelcast.nio.serialization.DefaultPortableWriter.writePortableArray(DefaultPortableWriter.java:159)
        at com.meinc.gameplay.domain.Question.writeAsPortableArray(Question.java:353)
        at com.meinc.gameplay.domain.Question.writePortable(Question.java:406)
        at com.hazelcast.nio.serialization.PortableSerializer.write(PortableSerializer.java:57)
        at com.hazelcast.nio.serialization.DefaultPortableWriter.writePortableArray(DefaultPortableWriter.java:163)
        at com.meinc.gameplay.domain.Event.writeAsPortableArray(Event.java:584)
        at com.meinc.gameplay.domain.Event.writePortable(Event.java:530)
        at com.hazelcast.nio.serialization.PortableSerializer.write(PortableSerializer.java:57)
        at com.hazelcast.nio.serialization.PortableSerializer.write(PortableSerializer.java:29)
        at com.hazelcast.nio.serialization.StreamSerializerAdapter.write(StreamSerializerAdapter.java:48)
        at com.hazelcast.nio.serialization.SerializationServiceImpl.toData(SerializationServiceImpl.java:176)
        ... 11 more

It appears that Portable objects are limited in size, which is not spelled out ANYWHERE that I can find the in Hazelcast documentation.

--- a/hazelcast/src/main/java/com/hazelcast/nio/serialization/DefaultPortableWriter.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/serialization/DefaultPortableWriter.java
@@ -156,7 +156,7 @@ public class DefaultPortableWriter implements PortableWriter {
         out.writeInt(len);
         if (len > 0) {
             final int offset = out.position();
-            out.position(offset + len * 4);
+            out.writeZeroBytes(len * 4);
             for (int i = 0; i < portables.length; i++) {
                 out.writeInt(offset + i * 4, out.position());
                 final Portable portable = portables[i];

--- a/hazelcast/src/test/java/com/hazelcast/nio/serialization/PortableTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/nio/serialization/PortableTest.java
@@ -308,6 +308,82 @@ public class PortableTest {
         }
     }
 
+    //https://github.com/hazelcast/hazelcast/issues/2172
+    @Test
+    public void test_issue2172_WritePortableArray() {
+        final SerializationService ss = new SerializationServiceBuilder().setInitialOutputBufferSize(16).build();
+
+        final TestObject2[] testObject2s = new TestObject2[100];
+        for (int i = 0; i < testObject2s.length; i++) {
+            testObject2s[i] = new TestObject2();
+        }
+        final TestObject1 testObject1 = new TestObject1(testObject2s);
+
+        ss.toData(testObject1);
+
+    }
+
+    class TestObject1 implements Portable {
+
+        private Portable[] portables;
+
+        public TestObject1() {
+        }
+
+        public TestObject1(Portable[] p) {
+            portables = p;
+        }
+
+        @Override
+        public int getFactoryId() {
+            return FACTORY_ID;
+        }
+
+        @Override
+        public int getClassId() {
+            return 1;
+        }
+
+        @Override
+        public void writePortable(PortableWriter writer) throws IOException {
+            writer.writePortableArray("list", portables);
+        }
+
+        @Override
+        public void readPortable(PortableReader reader) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    class TestObject2 implements Portable {
+
+        private String shortString;
+
+        public TestObject2() {
+            shortString = "Hello World";
+        }
+
+        @Override
+        public int getFactoryId() {
+            return FACTORY_ID;
+        }
+
+        @Override
+        public int getClassId() {
+            return 2;
+        }
+
+        @Override
+        public void writePortable(PortableWriter writer) throws IOException {
+            writer.writeUTF("shortString", shortString);
+        }
+
+        @Override
+        public void readPortable(PortableReader reader) throws IOException {
+            throw new UnsupportedOperationException();
+        }
+    }
+
     public static class TestPortableFactory implements PortableFactory {
 
         public Portable create(int classId) {

